File: AalsiTrader/ViewModels/PaperTradingViewModel.swift
import Foundation

@Observable
final class PaperTradingViewModel {
    var portfolio: PaperPortfolio?
    var trades: [PaperTrade] = []
    var metrics: PaperMetrics?
    var equityCurve: [EquityPoint] = []
    var approvals: [SigmaApproval] = []
    var mode: String = "paper"
    var autoTradingEnabled: Bool = false
    var requireSigmaApproval: Bool = true
    var modeEnabled: Bool = false

    var isLoading = false
    var error: String?
    var selectedTab: PaperTab = .portfolio

    var isLiveMode: Bool { mode == "live" }

    var openTrades: [PaperTrade] {
        trades.filter { $0.status == .open }
    }

    var closedTrades: [PaperTrade] {
        trades.filter { $0.status == .closed }
    }

    var pendingApprovals: [SigmaApproval] {
        approvals.filter { $0.status == .pending }
    }

    // MARK: - Data Loading

    func loadAll() async {
        isLoading = true
        error = nil
        do {
            async let p = PaperTradingService.fetchPortfolio(mode: mode)
            async let t = PaperTradingService.fetchTrades(mode: mode)
            async let m = PaperTradingService.fetchMetrics(mode: mode)
            async let e = PaperTradingService.fetchEquityCurve(mode: mode)
            async let a = PaperTradingService.fetchApprovals()

            let (portfolio, trades, metrics, curve, approvals) = try await (p, t, m, e, a)
            self.portfolio = portfolio
            self.trades = trades
            self.metrics = metrics
            self.equityCurve = curve
            self.approvals = approvals
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func approveSignal(_ tradeId: String) async {
        do {
            try await PaperTradingService.approveTradeSignal(tradeId)
            if let idx = approvals.firstIndex(where: { $0.tradeId == tradeId }) {
                approvals[idx].status = .approved
            }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func rejectSignal(_ tradeId: String) async {
        do {
            try await PaperTradingService.rejectTradeSignal(tradeId)
            if let idx = approvals.firstIndex(where: { $0.tradeId == tradeId }) {
                approvals[idx].status = .rejected
            }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    // MARK: - Mode Management

    func loadMode() async {
        do {
            let response = try await PaperTradingService.fetchMode()
            if let m = response.mode { self.mode = m }
            if let e = response.enabled { self.modeEnabled = e }
            if let a = response.autoTradingEnabled { self.autoTradingEnabled = a }
            if let s = response.requireSigmaApproval { self.requireSigmaApproval = s }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func setMode(_ newMode: String) async {
        do {
            let response = try await PaperTradingService.setMode(body: ["mode": newMode])
            if let m = response.mode { self.mode = m }
            if let e = response.enabled { self.modeEnabled = e }
            if let a = response.autoTradingEnabled { self.autoTradingEnabled = a }
            if let s = response.requireSigmaApproval { self.requireSigmaApproval = s }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func toggleAutoTrading() async {
        do {
            let newValue = !autoTradingEnabled
            let response = try await PaperTradingService.setMode(body: ["autoTradingEnabled": newValue])
            if let a = response.autoTradingEnabled { self.autoTradingEnabled = a }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func toggleSigmaApproval() async {
        do {
            let newValue = !requireSigmaApproval
            let response = try await PaperTradingService.setMode(body: ["requireSigmaApproval": newValue])
            if let s = response.requireSigmaApproval { self.requireSigmaApproval = s }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    // MARK: - WebSocket handlers

    func setupWebSocket() {
        let ws = WebSocketManager.shared
        ws.onPaperTradeOpen = { [weak self] trade in
            self?.trades.insert(trade, at: 0)
            if let p = self?.portfolio {
                var updated = p
                updated.openPositions += 1
                self?.portfolio = updated
            }
            HapticService.impact(.medium)
        }

        ws.onPaperTradeClose = { [weak self] trade in
            if let idx = self?.trades.firstIndex(where: { $0.id == trade.id }) {
                self?.trades[idx] = trade
            }
            HapticService.impact(.light)
        }

        ws.onPaperPortfolioUpdate = { [weak self] portfolio in
            self?.portfolio = portfolio
        }
    }
}

enum PaperTab: String, CaseIterable {
    case portfolio = "Portfolio"
    case trades = "Trades"
    case rules = "Rules"
    case analytics = "Analytics"
    case approvals = "Approvals"
}

File: AalsiTraderAndroid/app/src/main/java/com/aalsitrader/android/viewmodel/PaperTradingViewModel.kt
package com.aalsitrader.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aalsitrader.android.model.*
import com.aalsitrader.android.repository.PaperTradingRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class PaperTradingViewModel : ViewModel() {
    private val repository = PaperTradingRepository()

    private val _portfolio = MutableStateFlow<PaperPortfolio?>(null)
    val portfolio: StateFlow<PaperPortfolio?> = _portfolio.asStateFlow()

    private val _trades = MutableStateFlow<List<PaperTrade>>(emptyList())
    val trades: StateFlow<List<PaperTrade>> = _trades.asStateFlow()

    private val _metrics = MutableStateFlow<PaperMetrics?>(null)
    val metrics: StateFlow<PaperMetrics?> = _metrics.asStateFlow()

    private val _equityCurve = MutableStateFlow<List<EquityPoint>>(emptyList())
    val equityCurve: StateFlow<List<EquityPoint>> = _equityCurve.asStateFlow()

    private val _approvals = MutableStateFlow<List<SigmaApproval>>(emptyList())
    val approvals: StateFlow<List<SigmaApproval>> = _approvals.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _selectedTab = MutableStateFlow(0)
    val selectedTab: StateFlow<Int> = _selectedTab.asStateFlow()

    private val _mode = MutableStateFlow("paper")
    val mode: StateFlow<String> = _mode.asStateFlow()

    init {
        loadAll()
    }

    fun setTab(tab: Int) {
        _selectedTab.value = tab
    }

    fun loadAll() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            val m = _mode.value
            try {
                _portfolio.value = repository.getPortfolio(m)
            } catch (_: Exception) { }
            try {
                _trades.value = repository.getTrades(mode = m)
            } catch (_: Exception) { }
            try {
                _metrics.value = repository.getMetrics(m)
            } catch (_: Exception) { }
            try {
                _equityCurve.value = repository.getEquityCurve(mode = m)
            } catch (_: Exception) { }
            try {
                _approvals.value = repository.getApprovals()
            } catch (_: Exception) { }
            _isLoading.value = false
        }
    }

    fun approveTrade(tradeId: String) {
        viewModelScope.launch {
            try {
                repository.approveTrade(tradeId)
                _approvals.value = _approvals.value.map {
                    if (it.tradeId == tradeId) it.copy(status = ApprovalStatus.approved) else it
                }
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }

    fun rejectTrade(tradeId: String) {
        viewModelScope.launch {
            try {
                repository.rejectTrade(tradeId)
                _approvals.value = _approvals.value.map {
                    if (it.tradeId == tradeId) it.copy(status = ApprovalStatus.rejected) else it
                }
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }

    fun toggleMode() {
        val newMode = if (_mode.value == "paper") "live" else "paper"
        _mode.value = newMode
        loadAll()
    }

    fun clearError() {
        _error.value = null
    }
}
